### ISSUE TITLE
Incorrect error for locally scoped method which shadows trait method

### ISSUE THREAD

**Issue by bobrippling** [rust-analyzer/rust-analyzer#10936]
**rust-analyzer version**: rust-analyzer 04f03a360 2021-11-06 dev

**rustc version**: rustc 1.57.0-beta.2 (7e4c9eebd 2021-10-22)

---

rust-analyzer seems to pick `From::from` instead of the method defined on an locally scoped `impl` when checking the below method call, giving an error. `rustc` accepts all these examples without error.

```rust
fn mismatch() {
    struct A;

    impl A {
        fn from(_: i32, _: i32) -> Self {
            A
        }
    }

    let _a = A::from(1, 2); // error: expected 1 argument, found 2
}
```

Whereas this is fine:
```rust
struct A;

impl A {
    fn from(_: i32, _: i32) -> Self {
        A
    }
}

fn ok() {
    let _a = A::from(1, 2); // no error
}
```

We also see an error if just the `impl` block is locally scoped:
```rust
struct A;

fn mismatch() {
    impl A {
        fn from(_: i32, _: i32) -> Self {
            A
        }
    }

    let _a = A::from(1, 2); // error: expected 1 argument, found 2
}
```
